### PR TITLE
Fix sub asset repeated loading

### PR DESCRIPTION
**Pull request by zhuxudong**
## Summary by CodeRabbit

- **Bug Fixes**
  - Improved the URL generation logic for asset loading to ensure correct handling of sub-assets.

### CODE REVIEW

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]
This BUG is that the new Feat has broken the existing functions. This is a relatively serious incident and a unit test must be added to prevent similar situations from happening again!

File: packages/core/src/asset/ResourceManager.ts [hunk 1]
```diff
@@ -549,9 +589,14 @@ const rePropName = RegExp(
...
-  {
-    resolve: (value: any) => void;
-    reject: (reason: any) => void;
-  }
+  Record<
+    // sub asset url, ie. "textures[0]"
+    string,
+    {
+      resolve?: any;
+      reject?: any;
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:599]: Why use `any`?

File: packages/core/src/asset/ResourceManager.ts [hunk 2]
```diff
@@ -182,17 +182,48 @@ export class ResourceManager {
   /**
    * @internal
    */
-  _onSubAssetSuccess<T>(assetURL: string, value: T): void {
-    this._subAssetPromiseCallbacks[assetURL]?.resolve(value);
-    delete this._subAssetPromiseCallbacks[assetURL];
+  _onSubAssetSuccess<T>(assetBaseURL: string, assetSubPath: string, value: T): void {
+    const subPromise = this._subAssetPromiseCallbacks[assetBaseURL]?.[assetSubPath];
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:186]: Are you sure this is a Promise?

File: packages/core/src/asset/ResourceManager.ts [hunk 3]
```diff
@@ -182,17 +182,48 @@ export class ResourceManager {
...
   /**
    * @internal
    */
-  _onSubAssetFail(assetURL: string, value: (reason: any) => void): void {
-    this._subAssetPromiseCallbacks[assetURL]?.reject(value);
-    delete this._subAssetPromiseCallbacks[assetURL];
+  _onSubAssetFail(assetBaseURL: string, assetSubPath: string, value: (reason: any) => void): void {
+    const subPromise = this._subAssetPromiseCallbacks[assetBaseURL]?.[assetSubPath];
+    if (subPromise) {
+      subPromise.reject(value);
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:202]: `reject` maybe undefined

File: packages/core/src/asset/ResourceManager.ts [hunk 4]
```diff
@@ -182,17 +182,48 @@ export class ResourceManager {
...
+    } else {
+      (this._subAssetPromiseCallbacks[assetBaseURL] ||= {})[assetSubPath] = {
+        reject: value
+      };
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:206]: Are you sure???

File: packages/core/src/asset/ResourceManager.ts [hunk 5]
```diff
@@ -182,17 +182,48 @@ export class ResourceManager {
...
-  _onSubAssetSuccess<T>(assetURL: string, value: T): void {
-    this._subAssetPromiseCallbacks[assetURL]?.resolve(value);
-    delete this._subAssetPromiseCallbacks[assetURL];
+  _onSubAssetSuccess<T>(assetBaseURL: string, assetSubPath: string, value: T): void {
+    const subPromise = this._subAssetPromiseCallbacks[assetBaseURL]?.[assetSubPath];
+    if (subPromise) {
+      subPromise.resolve(value);
+    } else {
+      (this._subAssetPromiseCallbacks[assetBaseURL] ||= {})[assetSubPath] = {
+        resolve: value
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:191]: `subPromise.resolve(value)` will call,`  resolve: value` just assign，Are you writing unspoken rules again?

File: packages/core/src/asset/ResourceManager.ts [hunk 6]
```diff
@@ -182,17 +182,48 @@ export class ResourceManager {
...
+        resolve(subPromise.resolve);
+      } else if (subPromise?.reject) {
+        reject(subPromise.reject);
+      } else {
+        (this._subAssetPromiseCallbacks[assetBaseURL] ||= {})[assetSubPath] = {
+          resolve,
+          reject
+        };
+      }
+    });
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:226]: `_createSubAssetPromiseCallback`、`_onSubAssetSuccess`、`_onSubAssetFail`：The design of these three functions is confusing

File: packages/core/src/asset/ResourceManager.ts [hunk 7]
```diff
@@ -467,6 +500,13 @@ export class ResourceManager {
     return result;
   }
 
+  /**
+   * @internal
+   */
+  private _releaseSubAssetPromiseCallback(assetBaseURL: string): void {
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:506]: private add @internal?

File: packages/core/src/asset/ResourceManager.ts [hunk 8]
```diff
@@ -182,17 +182,48 @@ export class ResourceManager {
...
+    }
   }
 
   /**
    * @internal
    */
-  _onSubAssetFail(assetURL: string, value: (reason: any) => void): void {
-    this._subAssetPromiseCallbacks[assetURL]?.reject(value);
-    delete this._subAssetPromiseCallbacks[assetURL];
+  _onSubAssetFail(assetBaseURL: string, assetSubPath: string, value: (reason: any) => void): void {
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:199]: private

File: packages/core/src/asset/ResourceManager.ts [hunk 9]
```diff
@@ -182,17 +182,48 @@ export class ResourceManager {
...
+      (this._subAssetPromiseCallbacks[assetBaseURL] ||= {})[assetSubPath] = {
+        reject: value
+      };
+    }
+  }
+
+  /**
+   * @internal
+   */
+  _createSubAssetPromiseCallback<T>(assetBaseURL: string, assetSubPath: string): AssetPromise<T> {
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:213]: private

File: packages/core/src/asset/ResourceManager.ts [hunk 10]
```diff
@@ -393,21 +408,39 @@ export class ResourceManager {
         () => delete loadingPromises[assetURL]
       );
 
-      promise.catch((e) => {
-        this._onSubAssetFail(assetURL, e);
+      // Check whether load main asset
+      const mainPromise = loadingPromises[assetBaseURL] || this._loadMainAsset(loader, item, assetBaseURL);
+      mainPromise.catch((e) => {
+        this._onSubAssetFail(assetBaseURL, queryPath, e);
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:414]: Before logic is same?

**Comment by GuoLei1990** [comment]
Try opt Promise count

**Comment by GuoLei1990** [comment]
load sub texture(finished)，then load sub material has cache?

File: packages/core/src/asset/ResourceManager.ts [hunk 11]
```diff
@@ -182,17 +182,48 @@ export class ResourceManager {
   /**
    * @internal
    */
-  _onSubAssetSuccess<T>(assetURL: string, value: T): void {
-    this._subAssetPromiseCallbacks[assetURL]?.resolve(value);
-    delete this._subAssetPromiseCallbacks[assetURL];
+  _onSubAssetSuccess<T>(assetBaseURL: string, assetSubPath: string, value: T): void {
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:185]: assetSubPath has undefined

**Comment by GuoLei1990** [comment]
Create a issue with can't query sub asset array

File: packages/core/src/asset/ResourceManager.ts [hunk 12]
```diff
@@ -182,17 +182,48 @@ export class ResourceManager {
...
+      };
+    }
+  }
+
+  /**
+   * @internal
+   */
+  _createSubAssetPromiseCallback<T>(assetBaseURL: string, assetSubPath: string): AssetPromise<T> {
+    return new AssetPromise<T>((resolve, reject) => {
+      const subPromise = this._subAssetPromiseCallbacks[assetBaseURL]?.[assetSubPath];
```
Comment (GuoLei1990) [inline packages/core/src/asset/ResourceManager.ts:215]: This is not subPromise

File: packages/core/src/asset/ResourceManager.ts (hunk 8 above)
Comment (zhuxudong) [reply packages/core/src/asset/ResourceManager.ts:199]: it can invoked in other package

File: packages/core/src/asset/ResourceManager.ts (hunk 9 above)
Comment (zhuxudong) [reply packages/core/src/asset/ResourceManager.ts:213]: done

File: packages/core/src/asset/ResourceManager.ts (hunk 7 above)
Comment (zhuxudong) [reply packages/core/src/asset/ResourceManager.ts:506]: done

**Pull request merged**